src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+              perturbationSets.size() - 1)
+            );
+      }
+
+      // modify Ai (expand)
+      std::vector<size_t> Ae = Ai;
+      perturbationExpand(Ae);
+      if ( perturbationValid(Ae, numProj) )
+      {
+        perturbationSets.push_back(Ae); // add expanded set to sets
+        minHeap.push(
+            std::make_pair(
+              perturbationScore(Ae, scores),
+              perturbationSets.size() - 1)
+            );
+      }
+
+    }while (! perturbationValid(Ai, numProj)  );//Discard invalid perturbations
+
+    // add perturbation vector to probing sequence if valid
+    for (size_t i = 0; i < Ai.size(); ++i)
+      additionalProbingBins(positions(Ai[i]), pvec) += actions(Ai[i]);
+
+  }
+  */
+
+  // Perturbation sets (A) mark with 1 the (score, action, dimension) positions
+  // included in a given perturbation vector. Other spaces are 0.
+  std::vector<bool> Ao(2 * numProj);
+  Ao[0] = 1; // Smallest vector includes only smallest score.
+
+  std::vector< std::vector<bool> > perturbationSets;
+  perturbationSets.push_back(Ao); // Storage of perturbation sets.
+
+  std::priority_queue<
+    std::pair<double, size_t>,        // contents: pairs of (score, index)
+    std::vector<                      // container: vector of pairs
+      std::pair<double, size_t>
+      >,
+    std::greater< std::pair<double, size_t> > // comparator of pairs
+  > minHeap; // our minheap
+
+  // Start by adding the lowest scoring set to the minheap.
+  minHeap.push( std::make_pair(perturbationScore(Ao, scores), 0) );
+
+  // Loop invariable: after pvec iterations, additionalProbingBins contains pvec
+  // valid codes of the lowest-scoring bins (bins most likely to contain
+  // neighbors of the query).
+  for (size_t pvec = 0; pvec < T; ++pvec)
+  {
+    std::vector<bool> Ai;
+    do
+    {
+      // Get the perturbation set corresponding to the minimum score.
+      Ai = perturbationSets[ minHeap.top().second ];
+      minHeap.pop(); // .top() returns, .pop() removes
+
+      // Shift operation on Ai (replace max with max+1).
+      std::vector<bool> As = Ai;
+      if ( perturbationShift(As) && perturbationValid(As, numProj) )
+        // Don't add invalid sets.
+      {
+        perturbationSets.push_back(As); // add shifted set to sets
+        minHeap.push( 
+            std::make_pair(
+              perturbationScore(As, scores), 
+              perturbationSets.size() - 1)
+            );
+      }
+
+      // Expand operation on Ai (add max+1 to set).
+      std::vector<bool> Ae = Ai;
+      if ( perturbationExpand(Ae) && perturbationValid(Ae, numProj) ) 
+        // Don't add invalid sets.
+      {
+        perturbationSets.push_back(Ae); // add expanded set to sets
+        minHeap.push(
+            std::make_pair(
+              perturbationScore(Ae, scores),
+              perturbationSets.size() - 1)
+            );
+      }
+
+    }while (! perturbationValid(Ai, numProj)  );//Discard invalid perturbations
```

Comment:
This is weird spacing; should be `} while (!PerturbationValid(Ai, numProj)); // Discard invalid perturbations.`